### DISCUSSION TITLE
Profiling a Parser

### DISCUSSION THREAD

**Discussion by Julian** [tree-sitter/tree-sitter#994] [Q&A] [answered]
[My in-development-very-scrappy-parser-because-I-am-no-parsing-expert](https://github.com/Julian/tree-sitter-lean) now takes ~2 minutes to run `tree-sitter generate` on.

Any suggestions of what to run to figure out why that is?

**Comment by tek** [comment]
wow, two minutes :sweat_smile: I had the issue with haskell that it climbed from 4 to 12 seconds at some point; I would have liked a profiler there as well. very curious to see what would cause your generation time.

FYI in my case it was due to accidentally allowing some expressions in types.

**Reply by Julian** [reply]
I've been stealing ideas from your Haskell parser, which should eventually have significant overlap with mine I suspect. So if nothing else well done / thanks for doing that!

**Reply by tek** [reply]
thanks, and awesome that you could use it!

**Reply by Julian** [reply]
Oh and yeah I am reasonably sure my issue is exactly as you say, too many expressions allowed in places, specifically I think my way of doing function application is all wrong, I just would love to know *where* (and then I'll need to think about what to do about it...)

**Comment by tek** [comment]
one thing I noticed – your `sep0` helper allows the sequence to begin with a separator. no idea if this is legal in Lean, but if not, you might wanna move the `optional` to outside of the `seq`

**Reply by Julian** [reply]
Thanks! Good catch. It's not. [Fixed, after waiting the 2 minutes to test :D](https://github.com/Julian/tree-sitter-lean/commit/eb9a3c80049fa7b228935e8d83ceb7d65760e468). Appreciated!

**Comment by tek** [comment]
In lieu of a profiler, what I would do is to remove all entries from `command` or `primary_expression` and then adding them one by one (or bisect) to analyze what really boosts the generation time.
As a next step, also start inlining arbitrary symbols, this sometimes improves or worsens the parser size.

**Reply by Julian** [reply]
I tried that just before posting it and concluded... "most or many that I add back cause the slowdown" unfortunately. Like having 1 thing there compiles in `0.11s`, and then I put 2 and it jumps to `16s`. But maybe I should do so slower and more methodically.

The inlining I didn't try, but I guess I should as well.

(And appreciated again!)

**Reply by tek** [reply]
yeah in that case I would repeat the same thing with deeper nested symbols.

**Comment by maxbrunsfeld** [comment, marked as answer]
I'd love to put more work towards profiling parser-generation. I did add a `generate` flag called `--report-states-for-rule` that you call like this:

```
tree-sitter generate  --report-states-for-rule  primary_expression
```

It will dump all of the LR "item sets" (each of which corresponds to a parse state) where the given rule (e.g. `primary_expression`) is in-progress. If you just want an overall summary that shows a breakdown by rule name of how many states have that rule in-progress, then you can run it without the rule name as an argument, like this:

```
tree-sitter generate  --report-states-for-rule  -
```

The output is a bit noisy, but it did help me to understand why certain constructs required lots of states to parse.

**Reply by tek** [reply]
ooh awesome

**Comment by tek** [comment]
I looked at the grammar and the Lean manual a bit, and it seems to me that you use `_expression` very gratuitously – for example, the manual says "do notation cannot occur in an `if` condition", but there you allow any expression.
I could imagine that the generation time would grow massively because of this pattern even though the grammar is structurally fine.

Maybe separate some of those symbols out into a `_pure_expression`?

**Reply by Julian** [reply]
I'm sure you're correct, but in initially fiddling I couldn't get any faster whilst still passing my tests (I don't see that exact language in the Lean manual, though I'm not an expert in the language, but it's definitely a very very syntactically flexible language), will have to dig in more carefully. I *did* though slightly more methodically play around with removing terms and I see that removing `conditional`, `match`, and `lambda` from `_primary_expression` is a minimal set of removals that take me from ~`1m30s` down to `11s` for `tree-sitter generate`, so... will focus a bit on those, and on understanding what the CLI option @maxbrunsfeld shared tells me.

Appreciated again for the tips! Still learning (or relearning) how to do this kind of thing.

**Reply by Julian** [reply]
I guess an obvious commonality those 3 have in common is that they have multiple instances of `_expression` in them...

**Reply by tek** [reply]
I also landed on these three, but didn't have any success figuring out how to improve their performance. Changing `_expression` to `_primary_expression` in some of them made it even worse :shrug:

**Comment by Julian** [comment]
Gonna close this I suppose given the great tips so far and that presumably the summary is "what's there at the minute is that CLI flag, and beyond that I should basically play around with my specific grammar".

Appreciate the help!

**Reply by tek** [reply]
please keep reporting any progress!

**Reply by Julian** [reply]
Oh, very glad to if you're interested, you've definitely got relevant expertise from the Haskell stuff so your eyes are very valuable. Will let you know how I do...

**Reply by tek** [reply]
I'm also eager to learn more about improving performance!

**Reply by Julian** [reply]
I haven't spent a ton more time here yet, just slowly chipping at things, but one new thing I noticed is that `conditional`, which I defined as:

```
    conditional: $ => prec.right(1, seq(
      'if',
      $._expression,
      'then',
      $._expression,
      'else',
      $._expression,
    )),
```

if I remove the *middle* `_expression` in the `then` branch, cuts generate time by **3x**.

**Reply by tek** [reply]
what's your reasoning for using right associativity here?

**Reply by Julian** [reply]
The associativity is probably wrong there (I left a lot of them wrong waiting for me to add tests to force them to be correct essentially).

**Reply by tek** [reply]
maybe it would be more sensible to start working on soundness by writing lots of tests and commenting out all paths you're not working on to reduce generation time. I can't judge whether this would eventually solve the problem, but I can imagine that if the grammar is just very unsound (which is the case since you allow `expression` almost everywhere) and you solve conflicts by introducing precedences as band-aids, it might blow up the parser due to the many different variants that code can be parsed.

**Reply by tek** [reply]
just FYI I usually added an entry to `conflicts` whenever tree-sitter complained, and only started working with precedences when I was done writing all the rules and tests. My experience was that introducing associativity to solve conflicts caused complications when working on other parts of the grammar.

**Reply by Julian** [reply]
Thanks for the advice. I did precisely the opposite mostly because I wasn't sure which way was going to cause less pain (adding to `conflicts` or adding some `precedence` value).

Your experience is super helpful. Will give a shot at pushing things in that direction.

**Reply by tek** [reply]
I still don't know what exactly happens during generation when using a precedence, but I do know that conflicts delay the decisions to runtime, while precedences (at least if you're not using the new named precedences) have implications at generation time that may cause unwanted interactions with any rule that is in a choice situation with the precedented rule somhow.

**Reply by Julian** [reply]
My goal if I had any was precisely to get to use #939 which looked interesting. So what I was trying to do was to get all the precedence values to be positive numbers, which I couldn't manage to do. But as I've hopefully said before I don't know what I'm doing here :) just really trying to learn.

**Reply by tek** [reply]
yes, the named precedences are much easier to use!

**Comment by Julian** [comment]
Another tip someone finding this may find useful given I spent a few more minutes on this just now is... `git bisect` is just always your friend and you should always get over the laziness of using it :D -- in particular I ran essentially `git bisect run timeout 20 tree-sitter generate` (with some cleanup between executions and where `timeout` is from GNU coreutils) and it immediately spit out which commit of mine is the one responsible for making `tree-sitter generate` take more than 20 seconds.

In my case it's [this one](https://github.com/Julian/tree-sitter-lean/commit/c05bb9b1fa03a4ef083ba8684baf0bdad0c7d6e6) which in retrospect was super painful to get working so it's no wonder it introduced mistakes...

**Reply by tek** [reply]
did you figure out which part of that was the culprit?

**Reply by Julian** [reply]
Partially, but not in a very productive way yet. E.g., introducing this [`conditional_when` rule](https://github.com/Julian/tree-sitter-lean/commit/c05bb9b1fa03a4ef083ba8684baf0bdad0c7d6e6#diff-919ac210accac9ecc55a76d10a7590e3d85ca3f0e165b52d30f08faee486d0cbR403-R409) definitely contributes, it makes generation take a full additional minute, but I don't see how to remove it yet and still parse correctly.

**Reply by ahlinc** [reply]
Another finding is when I worked on a grammar I usually used my own a bit patched version of the `tree-sitter` CLI command that was build with `cargo b` and with every several new added rules I saw how the time for run the `tree-sitter generate` increases in seconds until the generation time came to a minute. But when I accidentally installed the latest release version of the `tree-sitter` command from a `tree-sitter-cli` npm package the generation time drops to less than 3 seconds. Then I've rebuild my patched version with `cargo b --release` and it shows the same less than 3 seconds time. So the build type of the `tree-sitter` cli tool may have a significant impact on the generation time.
